File: src/main/java/work/lclpnet/ap2/game/guess_it/data/GuessItManager.java
package work.lclpnet.ap2.game.guess_it.data;

import net.minecraft.server.world.ServerWorld;
import org.jetbrains.annotations.NotNull;
import work.lclpnet.ap2.api.game.MiniGameHandle;
import work.lclpnet.ap2.game.guess_it.challenge.*;
import work.lclpnet.ap2.game.guess_it.util.GuessItDisplay;
import work.lclpnet.ap2.impl.util.world.stage.Stage;
import work.lclpnet.lobby.util.WorldModifier;

import java.util.*;

public class GuessItManager {

    private final Random random;
    private final Set<Challenge> challenges = new LinkedHashSet<>();
    private final List<Challenge> queue = new ArrayList<>();

    public GuessItManager(MiniGameHandle gameHandle, ServerWorld world, Random random, Stage stage,
                          WorldModifier modifier, SoundSubtitles soundSubtitles) {
        this.random = random;

        var stageRadiusHeight = validateStage(stage);

        GuessItDisplay display = new GuessItDisplay(world, modifier, stage);

        challenges.add(new MathsChallenge(gameHandle, random));
        challenges.add(new DayTimeChallenge(gameHandle, world, random));
        challenges.add(new MobCountSingleChallenge(gameHandle, world, random, stage, modifier));
        challenges.add(new MobCountMultiChallenge(gameHandle, world, random, stage, modifier));
        challenges.add(new DistinctMobCountChallenge(gameHandle, world, random, stage, modifier));
        challenges.add(new SoundChallenge(gameHandle, world, random, soundSubtitles));
        challenges.add(new CakeBitesChallenge(gameHandle, world, random, stage, modifier));
        challenges.add(new PotionTypeChallenge(gameHandle, random, display));
        challenges.add(new FoodAmountChallenge(gameHandle, random, display));
        challenges.add(new ArmorTrimChallenge(gameHandle, world, random, stage, modifier));
        challenges.add(new BlockCountChallenge<>(gameHandle, random, stageRadiusHeight, modifier));
        challenges.add(new RecordChallenge(gameHandle, world, random, display));
        challenges.add(new AreaChallenge(gameHandle, world, random, stage, modifier));
        challenges.add(new MinecartChallenge(gameHandle, world, random, stage, modifier));
    }

    @SuppressWarnings("unchecked")
    private <S extends Stage & Stage.WithRadius & Stage.WithHeight> S validateStage(Stage stage) {
        if (!(stage instanceof Stage.WithRadius)) throw new IllegalArgumentException("Stage with radius is required");
        if (!(stage instanceof Stage.WithHeight)) throw new IllegalArgumentException("Stage with height is required");
        return (S) stage;
    }

    @NotNull
    public Challenge nextChallenge() {
        if (queue.isEmpty()) {
            if (challenges.isEmpty()) {
                throw new IllegalStateException("No challenges registered");
            }

            queue.addAll(challenges);
            Collections.shuffle(queue, random);
        }

        return queue.removeFirst();
    }
}

File: src/minigames/guess_it/src/main/kotlin/work/lclpnet/ap2/game/guess_it/data/GuessItManager.kt
package work.lclpnet.ap2.game.guess_it.data

import net.minecraft.server.level.ServerLevel
import work.lclpnet.ap2.game.MiniGameHandle
import work.lclpnet.ap2.game.guess_it.challenge.*
import work.lclpnet.ap2.game.guess_it.util.DynamicEntityModifier
import work.lclpnet.ap2.game.guess_it.util.GuessItDisplay
import work.lclpnet.ap2.impl.util.debug.DebugController
import work.lclpnet.ap2.impl.util.world.block_shape.BlockShape
import work.lclpnet.ap2.impl.util.world.block_shape.BlockShape.WithHeight
import work.lclpnet.ap2.impl.util.world.block_shape.BlockShape.WithRadius
import work.lclpnet.gaco.ds.IndexedSet
import work.lclpnet.game.util.WorldModifier
import java.util.*

class GuessItManager(
    gameHandle: MiniGameHandle,
    world: ServerLevel,
    private val random: Random,
    blockShape: BlockShape,
    modifier: WorldModifier,
    soundSubtitles: SoundSubtitles,
    debugController: DebugController,
    mannequinUuids: IndexedSet<UUID>,
    dynamicEntities: DynamicEntityModifier,
) {
    private val challenges = HashMap<String, Challenge>()
    private val queue = ArrayList<Challenge>()
    private val priority = ArrayList<ChallengeInit>()

    init {
        require(blockShape is WithRadius) { "Stage with radius is required" }
        require(blockShape is WithHeight) { "Stage with height is required" }

        val display = GuessItDisplay(world, modifier, blockShape)

        registerChallenge(MathsChallenge(gameHandle, random))
        registerChallenge(DayTimeChallenge(gameHandle, world, random, blockShape, dynamicEntities))
        registerChallenge(MobCountSingleChallenge(gameHandle, world, random, blockShape, modifier, mannequinUuids))
        registerChallenge(MobCountMultiChallenge(gameHandle, world, random, blockShape, modifier, mannequinUuids))
        registerChallenge(DistinctMobCountChallenge(gameHandle, world, random, blockShape, modifier, mannequinUuids))
        registerChallenge(SoundChallenge(gameHandle, world, random, soundSubtitles))
        registerChallenge(CakeBitesChallenge(gameHandle, world, random, blockShape, modifier, dynamicEntities))
        registerChallenge(PotionTypeChallenge(gameHandle, random, display))
        registerChallenge(FoodAmountChallenge(gameHandle, random, display))
        registerChallenge(ArmorTrimChallenge(gameHandle, world, random, blockShape, modifier))
        registerChallenge(
            BlockCountChallenge(
                gameHandle,
                random,
                blockShape,
                modifier,
                debugController
            )
        )
        registerChallenge(RecordChallenge(gameHandle, world, random, display))
        registerChallenge(AreaChallenge(gameHandle, world, random, blockShape, modifier))
        registerChallenge(MinecartChallenge(gameHandle, world, random, blockShape, modifier))
    }

    private fun registerChallenge(challenge: Challenge) {
        check(!challenges.containsKey(challenge.id())) { "Duplicate challenge id " + challenge.id() }

        challenges[challenge.id()] = challenge
    }

    fun nextChallenge(): ChallengeInit {
        if (!priority.isEmpty()) {
            return priority.removeFirst()
        }

        if (queue.isEmpty()) {
            check(!challenges.isEmpty()) { "No challenges registered" }

            queue.addAll(challenges.values)
            queue.shuffle(random)
        }

        return ChallengeInit(queue.removeFirst(), null)
    }

    fun pushChallenge(challenge: Challenge, init: Any?) {
        priority.add(ChallengeInit(challenge, init))
    }

    fun getChallenges(): Collection<Challenge> =
        challenges.values.toList()

    data class ChallengeInit(val challenge: Challenge, val init: Any?)
}
